docs/8-kubernetes-container-orchestration/8.7.1-kustomize.md

```diff
@@ -0,0 +1,277 @@
+```yaml
+  resources:
+    limits:
+      memory: 1Gi
+    requests:
+      memory: 512Mi
+      cpu: 100m
+```
+
+> Note that many documentation sources will have `patchesStrategicMerge` and `patchesJson6902` as Kustomize operators, but recent versions of Kustomize have simplified both to `patches`. The underlying usage of both hasn't changed, just what they're actually called in your `kustomization.yaml`.
+
+> Also note that some fields are immutable in Kubernetes. You may need to delete your resources and reapply. Delete by running `kustomize build . | kubectl delete -f -`
+
+## Exercise 4: Secret Management
+
+Often times, you need to create/apply secrets into your Kubernetes cluster so different containers can access them. While under the right circumstances, it's okay for the secrets to be visible locally and on your cluster (ie not completely hidden), it is never okay for them to be pushed to GitHub. Kustomize can help you structure your resources in a way that the secret *existence* will be visible in GitHub, but not the secret content itself.
+
+* Start off by creating a `.env` file and place a key,value pair inside of it. This file will only be stored locally, and should never be pushed to GitHub.
+
+`.env`
+```bash
+SECRET=123456
+```
+
+`.gitignore`
+```bash
+.env
+```
+
+* Now that your secret information is stored locally, look into the Kustomize `secretGenerator` to, well, generate a secret that contains your key,value pair.
+> Note that the secret will be created with base64 encoded information, so you'll need to decode it to actually read the contents of the secret.
+
+* Finally, write a Kustomize `patch` (either method from above) to add the following entry to your container's `env` list in the deployment:
+
+```yaml
+- name: SECRET
+  valueFrom:
+    secretKeyRef:
+      name: <The name you gave your secret>
+      key: SECRET
+```
+
+* Exec into your container running in the cluster and run `env` to confirm that your environment variable is present and holding the correct information.
+> Since you cannot change a containers base env list after booting, you may need to delete your deployment before applying this change.
+
+* Now that your secret exists, update your key,value pair in your `.env` file and reapply. Did your secret get updated? If so, good. If it didn't and instead a *new* secret was created with your change, figure out how to correct that.
+
+## Exercise 5: Script Embedding
+
+Earlier in the chapter you were shown the `ConfigMap` resource. As a refresher, a `ConfigMap` is a resource that holds *unstructured* data (basically whatever you want, stored as a string). ConfigMaps are commonly used to embed scripts into Kubernetes clusters, and are then mounted to containers for execution. However, the process of writing these scripts is painful, because you need to write them as one long string inside of a ConfigMap field.
```

Comment:
```suggestion
Earlier in the chapter, we introduced the ConfigMap resource. As a refresher, a ConfigMap is used to store unstructured data, typically in the form of key-value pairs. This data can include configuration files, environment variables, or any other external configuration you want to decouple from your container images. By externalizing configuration, you avoid baking it directly into the image, allowing for more flexibility and easier updates. ConfigMaps can also be used to inject configuration files or scripts into containers, supporting dynamic adjustments to application behavior without needing to rebuild images. However, the process of writing these configmaps can be painful because you need to write them as one long string inside of a ConfigMap field. Doing so IDEs will rightfully treat the data as just a string and you loose all the language sEarlier in the chapter, we introduced the ConfigMap resource. As a refresher, a ConfigMap is used to store unstructured data, typically in the form of key-value pairs. This data can include configuration files, environment variables, or any other external configuration you want to decouple from your container images. By externalizing configuration, you avoid baking it directly into the image, allowing for more flexibility and easier updates. ConfigMaps can also be used to inject configuration files or scripts into containers, supporting dynamic adjustments to application behavior without needing to rebuild images.

However, writing ConfigMaps can be cumbersome because you often need to embed data as a single long string within the ConfigMap fields. This approach causes IDEs to treat the data as plain text, resulting in the loss of helpful features like syntax highlighting, intellisense, and language-specific support that you would normally get when working with typed files (e.g., .sh, .css, .html). This makes managing and editing complex configuration data more challenging.
```